### ISSUE TITLE
Self is broken for generic async context managers

### ISSUE THREAD

**Issue by asleep-cult** [microsoft/pyright#2994]
**Describe the bug**
Self in a generic async context manager results in the base class when used in a subclasss.

**Screenshots or Code**
```python
import typing
from typing_extensions import Self

T = typing.TypeVar('T')

class Base(typing.Generic[T]):
    async def __aenter__(self) -> Self:
        return self

    async def __aexit__(self, *args: typing.Any) -> None:
        return None

class Subclass(Base[int]):
    ...

async def do():
    async with Subclass() as f:
        reveal_type(f)  # Type of "f" is "Self@Base[int]"
```

with a synchronous context manager
```py
import typing
from typing_extensions import Self

T = typing.TypeVar('T')

class Base(typing.Generic[T]):
    def __enter__(self) -> Self:
        return self

    def __exit__(self, *args: typing.Any) -> None:
        return None

class Subclass(Base[int]):
    ...

def do():
    with Subclass() as f:
        reveal_type(f)  # Type of "f" is "Subclass"
```

**VS Code extension or command-line**
pyright 1.1.218

**Additional context**
This is specific to async context managers that are generic. It can be fixed by annotating self with a typevar and using it as the return type.

**Comment by erictraut** [comment]
Thanks for the bug report. This will be fixed in the next release.

**Comment by erictraut** [comment]
This is addressed in pyright 1.1.219, which I just published. It will also be in the next release of pylance.
